components/customs/visualizations/best-fit-curve-line-chart.tsx

```diff
@@ -0,0 +1,135 @@
+import uniq from 'lodash/uniq';
+
+import { groupDataForRechartsOnce } from "./group-data-for-recharts/group-data-for-recharts-once";
+import { LineChart } from "./line-chart";
+import { useBestFitCurve } from "./line-fitting/use-best-fit-curve";
+import { LegendConfiguration } from "./stacked-bar-chart";
+import { generateRange } from '@/lib/utils';
+
+interface BestFitLineSettings {
+  maximumPolynomialOrder: number;
+  yAxisDomain: {
+    maximumValue: number;
+    minimumValue: number;
+  }
+  allowStrictlyIncreasingLinesOnly: boolean;
+}
+
+interface BestFitCurveLineChartProps<
+  TData extends Record<string, unknown>,
+  TPrimaryGroupingKey extends string,
+> {
+  graphId: string;
+  data: TData[];
+  primaryGroupingFunction: (data: TData) => TPrimaryGroupingKey | TPrimaryGroupingKey[];
+  primaryGroupingSortFunction?: (
+    a: TPrimaryGroupingKey,
+    b: TPrimaryGroupingKey
+  ) => number;
+  dataPointToXAxisValue: (input: { dataPoint: TData, primaryGroupingKey: TPrimaryGroupingKey }) => number;
+  xAxisValueToLabel: (input: { xAxisValue: number }) => string;
+  xAxisLabelSortingFunction?: (
+    a: string,
+    b: string
+  ) => number;
+  dataPointToYAxisValue: (input: { dataPoint: TData, primaryGroupingKey: TPrimaryGroupingKey }) => number;
+  formatYAxisValue: (input: { yAxisValue: number}) => number;
+  getLineColour: (input: { primaryGroupingKey: TPrimaryGroupingKey }) => string;
```

Comment:
I regret this whole xAxis/yAxis thing. Really this should be primary and secondary keys just like all of the other charts. Please bear with this for now, I'll do a cleanup PR where I make everything primary and secondary keys when all of this stuff is through and I have a clean slate to work with.